src/lib/swapper/swappers/ThorchainSwapper/ThorchainSwapper2.ts

```diff
@@ -0,0 +1,201 @@
+  GetTradeQuoteInput,
+  SwapErrorRight,
+  Swapper2,
+  TradeQuote,
+  TradeQuote2,
+  UnsignedTx,
+} from 'lib/swapper/api'
+import { assertUnreachable, evm } from 'lib/utils'
+import type { AccountMetadata } from 'state/slices/portfolioSlice/portfolioSliceCommon'
+import { isUtxoChainId } from 'state/slices/portfolioSlice/utils'
+import { selectFeeAssetById, selectUsdRateByAssetId } from 'state/slices/selectors'
+import { store } from 'state/store'
+
+import { getThorTradeQuote } from './getThorTradeQuote/getTradeQuote'
+import type { Rates, ThorChainId, ThorUtxoSupportedChainId } from './ThorchainSwapper'
+import { ThorchainSwapper } from './ThorchainSwapper'
+import { getSignTxFromQuote } from './utils/getSignTxFromQuote'
+
+// Gets a from address either
+// - derived from the input (for our own consumption with our AccountMetadata and ChainId structures)
+// - or simply falls the passed from address through, for external consumers
+type WithFromParams = {
+  chainId?: ChainId
+  accountMetadata?: AccountMetadata
+  wallet?: HDWallet
+  from?: string
+}
+
+export const withFrom =
+  <T, P>(wrappedFunction: (params: P & { from: string | undefined }) => Promise<T>) =>
+  async (params: P & WithFromParams): Promise<T> => {
+    const { chainId, accountMetadata, from: inputFrom, wallet } = params
+
+    let from: string | undefined
+    if (inputFrom) {
+      from = inputFrom
+    } else {
+      const chainAdapterManager = getChainAdapterManager()
+      if (!chainId) throw new Error('No chainId provided')
+      const adapter = chainAdapterManager.get(chainId) as ChainAdapter<ThorChainId>
+      if (!adapter) throw new Error(`No adapter for ChainId: ${chainId}`)
+      const accountNumber = accountMetadata?.bip44Params?.accountNumber
+      const accountType = accountMetadata?.accountType
+
+      from = await (async () => {
+        if (!wallet) throw new Error('Wallet required for adapter.getAddress() call')
+        if (!accountNumber) throw new Error('Account number required for adapter.getAddress() call')
+        if (!isUtxoChainId(chainId)) return adapter.getAddress({ wallet, accountNumber })
+        if (!accountType || !accountNumber) return undefined
+        const { xpub } = await (adapter as UtxoChainAdapter).getPublicKey(
+          wallet,
+          accountNumber,
+          accountType,
+        )
+        return xpub
+      })()
+    }
+
+    return wrappedFunction({ ...params, from })
+  }
+
+export const thorchain: Swapper2 = {
+  getTradeQuote: async (
+    input: GetTradeQuoteInput,
+    rates: Rates,
+  ): Promise<Result<TradeQuote2, SwapErrorRight>> => {
+    const tradeQuoteResult = await getThorTradeQuote(input, rates)
+
+    return tradeQuoteResult.map(tradeQuote => {
+      const { receiveAddress, affiliateBps } = input
+      const id = String(Date.now()) // TODO: get thorchain quote ID or use uuid
+
+      return { id, receiveAddress, affiliateBps, ...tradeQuote }
+    })
+  },
+
+  getUnsignedTx: withFrom(
```

Comment:
Keep the base implementation non-injected, only inject `isFrom(inner)` when calling

Reply:
Actually since this isn't consumed yet, will keep it as-is for now, @woodenfurniture FYI for the follow-up diff ^^